traffic_ops/traffic_ops_golang/trafficstats/trafficstats.go

```diff
@@ -0,0 +1,691 @@
+		c.Interval = DEFAULT_INTERVAL
+	} else if c.Interval = tc.TrafficStatsDurationFromString(interval); c.Interval == tc.InvalidDuration {
+		log.Errorf("Error parsing 'interval' query parameter: %v", e)
+		e = errors.New("Invalid interval!")
+		return
+	}
+
+	if ex, ok := i.Params["exclude"]; ok {
+		switch ex {
+		case "series":
+			c.ExcludeSeries = true
+		case "summary":
+			c.ExcludeSummary = true
+		default:
+			e = errors.New("Invalid exclude! Must be 'series' or 'summary'")
+			return
+		}
+	}
+
+	c.MetricType = i.Params["metricType"]
+	if _, ok := metricTypes[c.MetricType]; !ok {
+		e = fmt.Errorf("Unknown metric type: %s", c.MetricType)
+		return
+	}
+
+	var ok bool
+	if c.DeliveryService, ok = i.Params["deliveryServiceName"]; !ok {
+		if c.DeliveryService, ok = i.Params["deliveryService"]; !ok {
+			e = errors.New("You must specify deliveryService or deliveryServiceName!")
+			return
+		}
+
+		if dsID, err := strconv.ParseUint(c.DeliveryService, 10, 64); err != nil {
+			// sql.ErrNoRows does not *necessarily* mean the DS doesn't exist - an XMLID can simply
+			// be numeric, and so it was wrong to treat it as an ID in the first place.
+			xmlid := c.DeliveryService
+			if c.DeliveryService, err = getXMLIDFromID(dsID, i.Tx.Tx); err != nil && err != sql.ErrNoRows {
+				log.Errorf("Converting DSID to XMLID: %v", err)
+				e = errors.New("Internal Server Error")
+				code = http.StatusInternalServerError
+				return
+			} else if err == sql.ErrNoRows {
+				c.DeliveryService = xmlid
+			}
+		}
+	}
+
+	code = http.StatusOK
+	return
+}
+
+func GetDSStats(w http.ResponseWriter, r *http.Request) {
+	// Perl didn't require "interval", but it would only return summary data if it was not given
+	inf, userErr, sysErr, errCode := api.NewInfo(r, []string{"metricType", "startDate", "endDate"}, nil)
+	tx := inf.Tx.Tx
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, tx, errCode, userErr, sysErr)
+		return
+	}
+	defer inf.Close()
+
+	var c config
+	if c, userErr, errCode = configFromRequest(r, inf); userErr != nil {
+		sysErr = fmt.Errorf("Unable to process deliveryservice_stats request: %v", userErr)
+		api.HandleErr(w, r, tx, errCode, userErr, sysErr)
+		return
+	}
+
+	client, err := inf.CreateInfluxClient()
+	if err != nil {
+		if err == sql.ErrNoRows {
+			errCode = http.StatusServiceUnavailable
+			userErr = errors.New("No InfluxDB servers available!")
+			sysErr = userErr
+		} else {
+			errCode = http.StatusInternalServerError
+			userErr = nil
+			sysErr = err
+		}
+		api.HandleErr(w, r, tx, errCode, userErr, sysErr)
+		return
+	} else if client == nil {
+		userErr = errors.New("Traffic Stats is not configured!")
+		sysErr = userErr
+		errCode = http.StatusServiceUnavailable
+		api.HandleErr(w, r, tx, errCode, userErr, sysErr)
+		return
+	}
+	defer (*client).Close()
+
+	var dsTenant uint
+	if dsTenant, err = dsTenantIDFromXMLID(c.DeliveryService, tx); err != nil {
+		sysErr = err
+
+		if err == sql.ErrNoRows {
```

Comment:
I wonder if dsTenantIDFromXMLID should return `exists, dsTenant, err` instead of bubbling up a mysql error

Reply:
that would be more consistent with other code in the codebase.